lib/internal/bootstrap/loaders.js

```diff
@@ -137,166 +137,180 @@ let internalBinding;
   };
 }
 
-// Think of this as module.exports in this file even though it is not
-// written in CommonJS style.
-const loaderExports = {
-  internalBinding,
-  NativeModule,
-  require: nativeModuleRequire
-};
-
 const loaderId = 'internal/bootstrap/loaders';
-
-// Set up NativeModule.
-function NativeModule(id) {
-  this.filename = `${id}.js`;
-  this.id = id;
-  this.exports = {};
-  this.module = undefined;
-  this.exportKeys = undefined;
-  this.loaded = false;
-  this.loading = false;
-  this.canBeRequiredByUsers = !id.startsWith('internal/');
-}
-
-// To be called during pre-execution when --expose-internals is on.
-// Enables the user-land module loader to access internal modules.
-NativeModule.exposeInternals = function() {
-  for (const [id, mod] of NativeModule.map) {
-    // Do not expose this to user land even with --expose-internals.
-    if (id !== loaderId) {
-      mod.canBeRequiredByUsers = true;
-    }
-  }
-};
-
 const {
   moduleIds,
   compileFunction
 } = internalBinding('native_module');
 
-NativeModule.map = new Map();
-for (let i = 0; i < moduleIds.length; ++i) {
-  const id = moduleIds[i];
-  const mod = new NativeModule(id);
-  NativeModule.map.set(id, mod);
-}
+const getOwn = (target, property, receiver) => {
+  return ObjectPrototypeHasOwnProperty(target, property) ?
+    ReflectGet(target, property, receiver) :
+    undefined;
+};
 
-function nativeModuleRequire(id) {
-  if (id === loaderId) {
-    return loaderExports;
+/**
+ * An internal abstraction for the built-in JavaScript modules of Node.js.
+ * Be careful not to expose this to user land unless --expose-internals is
+ * used, in which case there is no compatibility guarantee about this class.
+ */
+class NativeModule {
+  /**
+   * A map from the module IDs to the module instances.
+   * @type {Map<string, NativeModule>}
+  */
+  static map = new Map(moduleIds.map((id) => [id, new NativeModule(id)]));
+
+  constructor(id) {
+    this.filename = `${id}.js`;
+    this.id = id;
+    this.canBeRequiredByUsers = !id.startsWith('internal/');
+
+    // The CJS exports object of the module.
+    this.exports = {};
```

Comment:
Wouldn't a public (or maybe private) field declaration be preferable? 

```js
class NativeModule {
  exports = {};
  #loaded = false;
  #loading = false;
  module = undefined;
  exportKeys = undefined;
}
```

I personally find it more readable than a declaration in the constructor, but that might just be me.

Reply:
i don't mind if private fields are used but i think we should keep field initialization in one spot (in this case, the constructor, since it assigns from arguments).